File: airline/flutter_app/lib/models/flight.dart
import 'package:flutter/material.dart';

enum FlightStatus {
  scheduled,
  boarding,
  departed,
  arrived,
  delayed,
  cancelled;

  static FlightStatus fromString(String status) {
    final s = status.toUpperCase();
    if (s.contains('РАСПИСАНИЮ') || s == 'SCHEDULED') {
      return FlightStatus.scheduled;
    } else if (s.contains('ЗАДЕРЖАН') || s == 'DELAYED') {
      return FlightStatus.delayed;
    } else if (s.contains('ОТМЕНЕН') || s.contains('ОТМЕНЁН') || s == 'CANCELLED') {
      return FlightStatus.cancelled;
    } else if (s.contains('ВЫЛЕТЕЛ') || s == 'DEPARTED' || s.contains('ПОЛЕТ') || s.contains('FLIGHT') || s.contains('AIR')) {
      return FlightStatus.departed;
    } else if (s.contains('ПРИБЫЛ') || s == 'ARRIVED' || s.contains('LANDED')) {
      return FlightStatus.arrived;
    } else if (s.contains('ПОСАДКА') || s == 'BOARDING') {
      return FlightStatus.boarding;
    }
    return FlightStatus.scheduled;
  }

  Color get color {
    switch (this) {
      case FlightStatus.scheduled:
        return Colors.green;
      case FlightStatus.boarding:
        return Colors.orange;
      case FlightStatus.departed:
        return Colors.blue;
      case FlightStatus.arrived:
        return Colors.grey;
      case FlightStatus.delayed:
        return Colors.red;
      case FlightStatus.cancelled:
        return Colors.black87;
    }
  }
}

class Flight {
  final int id;
  final String flightNumber;
  final String departureCity;
  final String arrivalCity;
  final DateTime departureTime;
  final DateTime arrivalTime;
  final String status; // Статус как строка из backend
  final int availableSeats;
  final int totalSeats;
  final double basePrice;
  final int aircraftId;
  final int originAirportId;
  final int destinationAirportId;
  final String? gate;
  final String terminal;

  Flight({
    required this.id,
    required this.flightNumber,
    required this.departureCity,
    required this.arrivalCity,
    required this.departureTime,
    required this.arrivalTime,
    required this.status,
    required this.availableSeats,
    required this.totalSeats,
    required this.basePrice,
    this.aircraftId = 0,
    this.originAirportId = 0,
    this.destinationAirportId = 0,
    this.gate,
    this.terminal = 'A',
  });

  factory Flight.fromJson(Map<String, dynamic>? json) {
    if (json == null) {
      return Flight.empty();
    }
    
    return Flight(
      id: json['id'] as int? ?? 0,
      flightNumber: (json['flight_number'] ?? 'N/A').toString(),
      departureCity: (json['departure_city'] ?? 'Unknown').toString(),
      arrivalCity: (json['arrival_city'] ?? 'Unknown').toString(),
      departureTime: _parseDate(json['departure_time']),
      arrivalTime: _parseDate(json['arrival_time']),
      status: (json['status'] ?? 'Scheduled').toString(),
      availableSeats: json['available_seats'] as int? ?? 0,
      totalSeats: json['total_seats'] as int? ?? 0,
      basePrice: (json['base_price'] as num? ?? 0).toDouble(),
      aircraftId: json['aircraft_id'] as int? ?? 0,
      originAirportId: json['origin_airport_id'] as int? ?? 0,
      destinationAirportId: json['destination_airport_id'] as int? ?? 0,
      gate: json['gate']?.toString(),
      terminal: (json['terminal'] ?? 'A').toString(),
    );
  }

  factory Flight.empty() {
    return Flight(
      id: 0,
      flightNumber: 'N/A',
      departureCity: 'Unknown',
      arrivalCity: 'Unknown',
      departureTime: DateTime.now(),
      arrivalTime: DateTime.now(),
      status: 'Scheduled',
      availableSeats: 0,
      totalSeats: 0,
      basePrice: 0.0,
    );
  }

  static DateTime _parseDate(dynamic date) {
    if (date == null) return DateTime.now();
    try {
      if (date is String && date.isNotEmpty) {
        return DateTime.parse(date);
      }
      return DateTime.now();
    } catch (_) {
      return DateTime.now();
    }
  }

  FlightStatus get statusEnum => FlightStatus.fromString(status);

  String get route => '$departureCity → $arrivalCity';
  String get duration {
    final duration = arrivalTime.difference(departureTime);
    final hours = duration.inHours;
    final minutes = duration.inMinutes % 60;
    return '${hours}ч ${minutes}м';
  }
}

class FlightDetail {
  final int id;
  final String flightNumber;
  final String departureCity;
  final String arrivalCity;
  final DateTime departureTime;
  final DateTime arrivalTime;
  final String status;
  final int availableSeats;
  final int totalSeats;
  final double basePrice;
  final String aircraftType;
  final String aircraftModel;
  final int aircraftId;
  final int originAirportId;
  final int destinationAirportId;
  final String? gate;
  final String terminal;
  final int durationMinutes;

  FlightDetail({
    required this.id,
    required this.flightNumber,
    required this.departureCity,
    required this.arrivalCity,
    required this.departureTime,
    required this.arrivalTime,
    required this.status,
    required this.availableSeats,
    required this.totalSeats,
    required this.basePrice,
    required this.aircraftType,
    required this.aircraftModel,
    required this.aircraftId,
    required this.originAirportId,
    required this.destinationAirportId,
    this.gate,
    required this.terminal,
    required this.durationMinutes,
  });

  factory FlightDetail.fromJson(Map<String, dynamic> json) {
    return FlightDetail(
      id: json['id'] ?? 0,
      flightNumber: json['flight_number'] ?? 'N/A',
      departureCity: json['departure_city'] ?? 'Unknown',
      arrivalCity: json['arrival_city'] ?? 'Unknown',
      departureTime: json['departure_time'] != null ? DateTime.parse(json['departure_time']) : DateTime.now(),
      arrivalTime: json['arrival_time'] != null ? DateTime.parse(json['arrival_time']) : DateTime.now(),
      status: json['status'] ?? 'Scheduled',
      availableSeats: json['available_seats'] ?? 0,
      totalSeats: json['total_seats'] ?? 0,
      basePrice: (json['base_price'] ?? 0).toDouble(),
      aircraftType: json['aircraft_type'] ?? 'Boeing 737',
      aircraftModel: json['aircraft_model'] ?? '737-800',
      aircraftId: json['aircraft_id'] ?? 0,
      originAirportId: json['origin_airport_id'] ?? 0,
      destinationAirportId: json['destination_airport_id'] ?? 0,
      gate: json['gate'],
      terminal: json['terminal'] ?? 'A',
      durationMinutes: json['duration_minutes'] ?? 0,
    );
  }

  FlightStatus get statusEnum => FlightStatus.fromString(status);

  String get route => '$departureCity → $arrivalCity';
  String get duration {
    final hours = durationMinutes ~/ 60;
    final minutes = durationMinutes % 60;
    return '${hours}ч ${minutes}м';
  }
}

File: airline/backend/app/models/flight.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Float, Enum as SQLEnum
from sqlalchemy.orm import relationship
from datetime import datetime
import enum
from app.core.database import Base


class FlightStatus(str, enum.Enum):
    SCHEDULED = "ПО РАСПИСАНИЮ"
    BOARDING = "ПОСАДКА"
    DEPARTED = "ВЫЛЕТЕЛ"
    ARRIVED = "ПРИБЫЛ"
    DELAYED = "ЗАДЕРЖАН"
    CANCELLED = "ОТМЕНЕН"


class Flight(Base):
    __tablename__ = "flights"

    id = Column(Integer, primary_key=True, index=True)
    flight_number = Column(String, unique=True, index=True, nullable=False)
    aircraft_id = Column(Integer, ForeignKey("aircrafts.id"), nullable=True)
    origin_airport_id = Column(Integer, ForeignKey("airports.id"), nullable=False)
    destination_airport_id = Column(Integer, ForeignKey("airports.id"), nullable=False)
    scheduled_departure = Column(DateTime, nullable=False)
    scheduled_arrival = Column(DateTime, nullable=False)
    actual_departure = Column(DateTime, nullable=True)
    actual_arrival = Column(DateTime, nullable=True)
    status = Column(SQLEnum(FlightStatus), default=FlightStatus.SCHEDULED, nullable=False)
    base_price = Column(Float, nullable=False)
    gate = Column(String, nullable=True)
    terminal = Column(String, default="A", nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    aircraft = relationship("Aircraft", back_populates="flights")
    origin_airport = relationship("Airport", foreign_keys=[origin_airport_id], back_populates="origin_flights")
    destination_airport = relationship("Airport", foreign_keys=[destination_airport_id], back_populates="destination_flights")
    bookings = relationship("Booking", back_populates="flight", cascade="all, delete-orphan")
    seat_holds = relationship("SeatHold", back_populates="flight", cascade="all, delete-orphan")
    tickets = relationship("Ticket", back_populates="flight", cascade="all, delete-orphan")
    announcements = relationship("Announcement", back_populates="flight", cascade="all, delete-orphan")

    @property
    def departure_city(self):
        try:
            return self.origin_airport.city if self.origin_airport else "Unknown"
        except Exception:
            return "Unknown"

    @property
    def arrival_city(self):
        try:
            return self.destination_airport.city if self.destination_airport else "Unknown"
        except Exception:
            return "Unknown"

    @property
    def departure_time(self):
        return self.scheduled_departure

    @property
    def arrival_time(self):
        return self.scheduled_arrival

    @property
    def total_seats(self):
        if self.aircraft:
            return self.aircraft.capacity
        return 0

    @property
    def available_seats(self):
        total = self.total_seats
        if total <= 0:
            return 0
        
        try:
            from app.models.booking import BookingStatus
            # Считаем подтверждённые и ожидающие оплаты бронирования
            # Используем генератор для эффективности и проверяем наличие атрибутов
            occupied = sum(1 for b in self.bookings if b.status in [BookingStatus.CONFIRMED, BookingStatus.CREATED])
            return max(0, total - occupied)
        except Exception:
            # Если возникла любая ошибка (например, DetachedInstanceError при ленивой загрузке),
            # возвращаем общее число мест как запасной вариант.
            return total

    @property
    def aircraft_type(self):
        return "Boeing 737"

    @property
    def aircraft_model(self):
        if self.aircraft:
            return self.aircraft.model
        return "Unknown"

    @property
    def duration_minutes(self):
        if self.scheduled_arrival and self.scheduled_departure:
            delta = self.scheduled_arrival - self.scheduled_departure
            return int(delta.total_seconds() / 60)
        return 0



